toplevel/identity/entity.go

```diff
@@ -0,0 +1,642 @@
+	}
+	if existingEntities != nil {
+		getExistingEntitiesDetails(existingEntities, threadPoolSize, customMetadataSupported)
+		if !customMetadataSupported {
+			pruneUnsupported(desired)
+		}
+		populateAliasType(existingEntities)
+		copyIds(desired, existingEntities)
+		sortPolicies(existingEntities)
+	}
+
+	// determine entity changes
+	entitiesToBeWritten, entitiesToBeDeleted, entitiesToBeUpdated :=
+		vault.DiffItems(entriesAsItems(desired), entriesAsItems(existingEntities))
+	// determine entity alias changes
+	aliasesToBeWritten, aliasesToBeDeleted, aliasesToBeUpdated :=
+		determineAliasActions(desired, existingEntities, entitiesToBeDeleted)
+
+	// preform actions
+	if dryRun {
+		entitiesDryRunOutput(entitiesToBeWritten, "written")
+		entitiesDryRunOutput(entitiesToBeDeleted, "deleted")
+		entitiesDryRunOutput(entitiesToBeUpdated, "updated")
+		aliasesDryRunOutput(aliasesToBeWritten["id"], "written")
+		aliasesDryRunOutput(aliasesToBeWritten["name"], "written")
+		for _, alias := range aliasesToBeDeleted {
+			log.WithField("name", alias.Key()).WithField("type", alias.(entityAlias).AuthType).Info(
+				fmt.Sprintf("[Dry Run] [Vault Identity] entity alias to be delelted"))
+		}
+		aliasesDryRunOutput(aliasesToBeUpdated, "updated")
+	} else {
+		// TODO: make each action perform concurrently
+		for _, w := range entitiesToBeWritten {
+			w.(entity).CreateOrUpdate("written")
+		}
+		for _, d := range entitiesToBeDeleted {
+			d.(entity).Delete()
+		}
+		for _, u := range entitiesToBeUpdated {
+			u.(entity).CreateOrUpdate("update")
+		}
+		err = performAliasReconcile(aliasesToBeWritten, aliasesToBeDeleted, aliasesToBeUpdated, customMetadataSupported)
+		if err != nil {
+			log.WithError(err).Fatal(
+				"[Vault Identity] failed to perform entity-alias reconcile operations")
+		}
+	}
+}
+
+// generatedDesired accepts the yaml-marshalled result of the `vault_entities` graphql
+// query and returns entity/entity-alias objects
+func generateDesired(entries []user) []entity {
+	desired := []entity{}
+
+	for _, entry := range entries {
+		var newDesired *entity
+		var uniquePolicies map[string]bool
+		if entry.Roles != nil {
+			for _, role := range entry.Roles {
+				if role.Permissions != nil {
+					for _, permission := range role.Permissions {
+						if permission.Service == "vault" {
+							if newDesired == nil {
+								newDesired = &entity{
+									Name: entry.OrgUsername,
+									Type: "entity",
+									Aliases: []entityAlias{
+										{
+											Name:     entry.OrgUsername,
+											Type:     "entity-alias",
+											AuthType: "oidc",
+											CustomMetadata: map[string]interface{}{
+												"account": "RH SSO", // revisit this for better info to add
+											},
+										},
+									},
+									Metadata: map[string]interface{}{
+										"name": entry.Name,
+									},
+								}
+								uniquePolicies = make(map[string]bool)
+							}
+							for _, policy := range permission.Policies {
```

Comment:
I'm curious why policies are managed in this block. Isn't there another part of vault-manager which handles policies?

Reply:
Does this have the side effect of writing policies?

Reply:
If policies are included just to create entities and entity aliases for users who have roles, this is OK. If we are also managing and writing policies in this block, we should re-work the approach.